### DISCUSSION TITLE
How do I add a custom formatter in ASP.NET and AutoFac

### DISCUSSION THREAD

**Discussion by MorganBard** [axuno/SmartFormat#421] [Q&A] [answered]
Greetings all,

I have created a custom formatter and tried to add it in the Startup routine of an ASP.Net application. Unfortunately, it does not seem to be available later when I try to use it. I get an exception that it can't find the formatter. I worked around the problem by calling 

`Smart.Default.AddExtensions(formatter);`

In the constructor of the classes where I use it. This works, but seems counter-intuitive. Is this the way it's supposed to work or am I doing something wrong?

My thanks :)

Morgan Haldane

**Comment by axunonb** [comment]
Yes, this does not work
> `Smart.Default.AddExtensions(formatter);`

because the custom extension is registered in a `[ThreadStatic]` context, and thus an isolated version of the `SmartFormatter` is created in your *current* thread.

In ASP.NET you could create a new `SmartFormatter` instance with all extensions you need, and register it as service with `IServiceCollection.AddTransient(...)`. This way you can inject it where needed. Does this make sense to you?

**Reply by axunonb** [reply]
@MorganBard  Is this answer understandable?

**Reply by MorganBard** [reply]
My apologies. Life got super busy over the last few weeks and I had forgotten I posted this. That was rather rude, and you have my apologies.

I do understand the answer in concept. I don't use the built in ASP.Net dependency injection, though. I use AutoFac. Do you know how it would work with that DI framework?

Thanks very much!
Morgan

**Reply by axunonb** [reply]
For AutoFac the code would be similar to this (untested):
```CSharp
var builder = new ContainerBuilder();
builder.Register(c =>
{
    var formatter = new SmartFormatter();
    // Modify the formatter instance to match your needs
    return formatter;
}).InstancePerDependency();
```

**Reply by MorganBard** [reply]
I added a new answer with what I figured out. I had one more question, though. Is there a benefit to using the `InstancePerDependency` as opposed to `SingleInstance`? I was thinking the singleton would save some time since I call this fairly often in my code. Is there an issue with doing it as a singleton?

Thanks again,

Morgan

PS. This is great stuff! Thanks for putting it out there for folks to use!

**Reply by axunonb** [reply]
`SmartFormatter` is not thread-safe, that's why `SingleInstance` does not work. The time to create the formatter is negligible in an asp.net context.

**Reply by MorganBard** [reply]
Good to know! Thanks again. 

Morgan

**Comment by MorganBard** [comment, marked as answer]
Ok, after a bit of research, I got this and it seems to work:

In my ApplicationModule class (where I do all my dependency injection registrations) i have this in the Load method:

```c#
builder.Register(c =>
{
   SmartFormatter formatter = new SmartFormatter()
      .AddExtensions(
         new StringSource(),
         // will automatically be added to the IFormatter list, too
         new ListFormatter(),
         new DictionarySource(),
         new ValueTupleSource(),
         new ReflectionSource(),
         // for string.Format behavior
         new DefaultSource(),
         new KeyValuePairSource())
      .AddExtensions(
         new PluralLocalizationFormatter(),
         new ConditionalFormatter(),
         new IsMatchFormatter(),
         new NullFormatter(),
         new ChooseFormatter(),
         new SubStringFormatter(),
         // for string.Format behavior
         new DefaultFormatter());

   IFormatter numberFormatter = c.ResolveNamed<IFormatter>("nbrFormatter");
   formatter.AddExtensions(numberFormatter);

   return formatter;
}).AsSelf().InstancePerDependency();
```
In the places where I need formatting, I have this:

```c#
private readonly SmartFormatter formatter;

public class WorkerClass
{
   // Other constructor stuff here...
   formatter = factory.Create<SmartFormatter>(); // factory is my factory class used to create objects. It uses AutoFac internally.
}

public void WorkerMethod()
{
   int someValue;
   // method does stuff
   string stuff = formatter.Format("{value:nbrFormatter(x)}", new {value = someValue});
   DoSomethingWithStuff(stuff);
}
```

Initially, I got an error saying I was missing SourceExtensions. I looked up how the default formatter was initialized and copied that code here. The Wiki didn't have anything that I could find, other than a warning to "for the beginning, leave `SmartFormatter.Format()` alone." I figured it was a more advanced topic, but I couldn't find it anywhere else. So, a little figuring, some searches in the code here and I discovered how to do it. It works just fine, so I thought I'd include it here as a suggested answer for others to find if they needed it. 

Thanks,

Morgan

**Reply by axunonb** [reply]
Adding all extensions can be abbreviated with `var formatter = Smart.CreateDefaultSmartFormat()`, and then add your customer formatter with `formatter.AddExtensions(numberFormatter)`. For improved performance you might consider to only load the extensions that you actually need.

**Reply by MorganBard** [reply]
Also good to know! Thanks. Might it be possible to add something about using non-default formatter to the Wiki? If it's editable, I'd be happy add it myself.

Morgan

**Reply by axunonb** [reply]
Yes, the Wiki is editable. Your welcome to contribute.
